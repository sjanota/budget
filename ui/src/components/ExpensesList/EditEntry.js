import React from 'react';
import PropTypes from 'prop-types';
import { Expense } from '../../model/propTypes';
import * as MoneyAmount from '../../model/MoneyAmount';

export function EditEntry({ entry, setEntry }) {
  function setValue(value) {
    return setEntry(e => ({ ...e, ...value }));
  }
  return (
    <>
      <td>
        <input
          value={entry.title}
          onChange={event => setValue({ title: event.target.value })}
          type={'text'}
        />
      </td>
      <td>
        <input
          value={entry.date || ''}
          onChange={event => setValue({ date: event.target.value })}
          type={'date'}
        />
      </td>
      <td>
        <input
          value={MoneyAmount.format(entry.totalBalance)}
          onChange={event =>
            setValue({ totalBalance: MoneyAmount.parse(event.target.value) })
          }
          type={'number'}
        />
      </td>
      <td>
        <input
          value={entry.location || ''}
          onChange={event => setValue({ location: event.target.value })}
          type={'text'}
        />
      </td>
      <td />
    </>
  );
}

EditEntry.propTypes = {
  entry: Expense.isRequired,
  setEntry: PropTypes.func.isRequired,
};
