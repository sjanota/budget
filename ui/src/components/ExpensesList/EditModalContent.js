import React, { useState, useRef, useEffect } from 'react';
import { shape, string, arrayOf, any, func, number } from 'prop-types';
import { replaceOnList } from '../../util/immutable';
import { Modal, Form, Button, Row, Col } from 'react-bootstrap';
import { CreateButton } from '../common/CreateButton';
import * as MoneyAmounts from '../../model/MoneyAmount';
import { MoneyAmount, Expense } from '../../model/propTypes';
import { useQuery } from '@apollo/react-hooks';
import { useBudget } from '../context/budget/budget';
import { QUERY_CATEGORIES } from '../CategoriesList/CategoriesList.gql';
import { StateEntry } from './EditModalContent.propsTypes';

export function EditModalContent({ init, onCancel, onSubmit, autoFocusRef }) {
  const [state, setState] = useState(init);
  const titleRef = useRef();

  useEffect(() => {
    if (titleRef.current) titleRef.focus();
  }, [titleRef]);

  function setValue(value) {
    return setState(e => ({ ...e, ...value }));
  }

  const inputProps = {
    state,
    setValue,
  };

  return (
    <>
      <Modal.Header>Nowy wydatek</Modal.Header>
      <Modal.Body>
        <Form>
          <Row>
            <TitleInput {...inputProps} autoFocusRef={autoFocusRef} />
          </Row>
          <Row>
            <DateInput {...inputProps} />
            <AccountInput {...inputProps} />
          </Row>
          <hr />
          <CategoriesList state={state} setState={setState} />
          <hr />
          <Row>
            <SumOutput state={state} />
          </Row>
        </Form>
      </Modal.Body>
      <Modal.Footer>
        <CancelButton onCancel={onCancel} />
        <SubmitButton state={state} onSubmit={onSubmit} onCancel={onCancel} />
      </Modal.Footer>
    </>
  );
}
EditModalContent.propTypes = {
  init: Expense,
  onCancel: func.isRequired,
  onSubmit: func.isRequired,
  autoFocusRef: any,
};

function TitleInput({ state, autoFocusRef, setValue }) {
  return (
    <Col>
      <Form.Label>Tytuł</Form.Label>
      <Form.Control
        ref={autoFocusRef}
        type="text"
        placeholder="Tytuł"
        value={state.title}
        onChange={e => setValue({ title: e.target.value })}
      />
    </Col>
  );
}
TitleInput.propTypes = {
  state: shape({ title: string.isRequired }),
  setValue: func.isRequired,
  autoFocusRef: EditModalContent.autoFocusRef,
};

function DateInput({ state, setValue }) {
  return (
    <Col>
      <Form.Label>Data</Form.Label>
      <Form.Control
        type="date"
        placeholder="Data"
        value={state.date}
        onChange={e => setValue({ date: e.target.value })}
      />
    </Col>
  );
}
DateInput.propTypes = {
  state: shape({ date: string.isRequired }),
  setValue: func.isRequired,
};

function AccountInput() {
  return (
    <Col>
      <Form.Label>Konto</Form.Label>
      <Form.Control
        type="text"
        placeholder="Under construction"
        readOnly={true}
      />
    </Col>
  );
}
AccountInput.propTypes = {};

function CreateCategoryButton({ setState }) {
  return (
    <CreateButton
      onClick={() =>
        setState(s => ({
          ...s,
          entries: [
            ...s.entries,
            { title: '', balance: '0.0', categoryID: '' },
          ],
        }))
      }
    />
  );
}
CreateCategoryButton.propTypes = {
  setState: func.isRequired,
};

function CategoriesList({ state, setState }) {
  function setEntry(idx, update) {
    return setState(s => {
      const entries = replaceOnList(s.entries, idx, {
        ...s.entries[idx],
        ...update,
      });
      const totalBalance = entries.reduce(
        (acc, v) => MoneyAmounts.add(acc, v.balance),
        MoneyAmounts.zero()
      );
      return {
        ...s,
        entries,
        totalBalance,
      };
    });
  }

  return (
    <>
      <Form.Label>Kategorie</Form.Label>
      <CreateCategoryButton setState={setState} />
      {state.entries.map((entry, idx) => (
        <Row key={idx}>
          <CategoryRow entry={entry} idx={idx} setEntry={setEntry} />
        </Row>
      ))}
    </>
  );
}
CategoriesList.propTypes = {
  setState: func.isRequired,
  state: shape({
    entries: arrayOf(StateEntry),
  }),
};

function SumOutput({ state }) {
  return (
    <>
      <Col>
        <Form.Label>Suma</Form.Label>
      </Col>
      <Col>
        <Form.Control
          type="number"
          placeholder="Suma"
          value={MoneyAmounts.format(state.totalBalance)}
          readOnly={true}
        />
      </Col>
    </>
  );
}
SumOutput.propTypes = {
  state: shape({ totalBalance: MoneyAmount }),
};

function CancelButton({ onCancel }) {
  return (
    <Button variant="secondary" onClick={onCancel}>
      Anuluj
    </Button>
  );
}
CancelButton.propTypes = {
  onCancel: EditModalContent.onCancel,
};

function SubmitButton({ state, onSubmit, onCancel }) {
  return (
    <Button
      variant="primary"
      onClick={() => {
        onSubmit(state);
        onCancel();
      }}
    >
      Zapisz
    </Button>
  );
}
SubmitButton.propTypes = {
  state: any.isRequired,
  onSubmit: EditModalContent.onSubmit,
  onCancel: EditModalContent.onCancel,
};

function CategoryRow(props) {
  return (
    <>
      <Col>
        <CategoryInput {...props} />
      </Col>
      <Col>
        <CategoryBalanceInput {...props} />
      </Col>
    </>
  );
}
CategoryRow.propTypes = {
  entry: StateEntry,
  idx: number.isRequired,
  setEntry: func.isRequired,
};

function CategoryInput({ entry, idx, setEntry }) {
  const { id: budgetID } = useBudget();
  const { loading, error, data } = useQuery(QUERY_CATEGORIES, {
    variables: { budgetID },
  });
  if (loading) return <p>Loading...</p>;
  if (error) {
    console.error(error);
    return <p>Error :(</p>;
  }

  return (
    <Form.Control
      as="select"
      value={entry.categoryID || (entry.category && entry.category.id)}
      onChange={e => setEntry(idx, { categoryID: e.target.value })}
    >
      <option></option>
      {data.categories
        .sort((c1, c2) => c1.name.localeCompare(c2))
        .map(category => (
          <option key={category.id} value={category.id}>
            {category.name}
          </option>
        ))}
    </Form.Control>
  );
}
CategoryInput.propTypes = {
  entry: StateEntry,
  idx: number.isRequired,
  setEntry: func.isRequired,
};

function CategoryBalanceInput({ entry, idx, setEntry }) {
  return (
    <Form.Control
      type="number"
      placeholder="Kwota"
      value={MoneyAmounts.format(entry.balance)}
      onChange={e => setEntry(idx, { balance: e.target.value })}
      onBlur={() =>
        setEntry(idx, { balance: MoneyAmounts.parse(entry.balance) })
      }
    />
  );
}
CategoryBalanceInput.propTypes = {
  entry: StateEntry,
  idx: number.isRequired,
  setEntry: func.isRequired,
};
