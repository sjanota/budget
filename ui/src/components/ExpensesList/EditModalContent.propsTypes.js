import { shape, string, any, oneOfType } from 'prop-types';
import { MoneyAmount } from '../../model/propTypes';

export const StateEntry = shape({
  categoryID: any,
  category: shape({
    id: any,
  }),
  balance: oneOfType([string, MoneyAmount]),
});
