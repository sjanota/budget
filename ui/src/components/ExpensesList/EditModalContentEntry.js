import React from 'react';
import PropTypes from 'prop-types';
import { useQuery } from '@apollo/react-hooks';
import { useBudget } from '../context/budget/budget';
import { Form, Row, Col } from 'react-bootstrap';
import { QUERY_CATEGORIES } from '../CategoriesList/CategoriesList.gql';
import * as MoneyAmount from '../../model/MoneyAmount';
import { MoneyAmount as MoneyAmountPropType } from '../../model/propTypes';

const propTypes = {
  entry: PropTypes.shape({
    categoryID: PropTypes.any,
    category: PropTypes.shape({
      id: PropTypes.any.isRequired,
    }),
    balance: PropTypes.oneOfType([PropTypes.string, MoneyAmountPropType]),
  }),
  idx: PropTypes.number.isRequired,
  setEntry: PropTypes.func.isRequired,
};

export function EditModalContentEntry(props) {
  return (
    <Row>
      <Col>
        <CategoryInput {...props} />
      </Col>
      <Col>
        <BalanceInput {...props} />
      </Col>
    </Row>
  );
}
EditModalContentEntry.propTypes = propTypes;

function CategoryInput({ entry, idx, setEntry }) {
  const { id: budgetID } = useBudget();
  const { loading, error, data } = useQuery(QUERY_CATEGORIES, {
    variables: { budgetID },
  });
  if (loading) return <p>Loading...</p>;
  if (error) {
    console.error(error);
    return <p>Error :(</p>;
  }

  return (
    <Form.Control
      as="select"
      value={entry.categoryID || (entry.category && entry.category.id)}
      onChange={e => setEntry(idx, { categoryID: e.target.value })}
    >
      <option></option>
      {data.categories
        .sort((c1, c2) => c1.name.localeCompare(c2))
        .map(category => (
          <option key={category.id} value={category.id}>
            {category.name}
          </option>
        ))}
    </Form.Control>
  );
}

CategoryInput.propTypes = propTypes;

function BalanceInput({ entry, idx, setEntry }) {
  return (
    <Form.Control
      type="number"
      placeholder="Kwota"
      value={MoneyAmount.format(entry.balance)}
      onChange={e => setEntry(idx, { balance: e.target.value })}
      onBlur={() =>
        setEntry(idx, { balance: MoneyAmount.parse(entry.balance) })
      }
    />
  );
}

BalanceInput.propTypes = propTypes;
