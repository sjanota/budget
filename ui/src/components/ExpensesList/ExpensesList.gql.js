import gql from "graphql-tag";

const EXPENSE_FRAGMENT = gql`
    fragment ExpensesDetails on Expense {
        id
        title
        date
        total {
            integer
            decimal
        }
        location
        account {
            id
            name
        }
    }
`;

export const EXPENSES_QUERY = gql`
    query QueryExpenses {
        expenses {
            ...ExpensesDetails
        }
    }
    ${EXPENSE_FRAGMENT}
`;

export const EXPENSES_EVENTS_SUBSCRIPTION = gql`
    subscription WatchExpenses {
        expenseEvents {
            type
            expense {
                ...ExpensesDetails
            }
        }
    }
    ${EXPENSE_FRAGMENT}
`;

export const DELETE_EXPENSE = gql`
  mutation DeleteExpense($id: ID!) {
      deleteExpense(id: $id) {
          id
      }
  }
`;

export const UPDATE_EXPENSE = gql`
    mutation UpdateExpense($id: ID!, $input: ExpenseInput!) {
        updateExpense(id: $id, input: $input) {
            id
        }
    }
`;

export const CREATE_EXPENSE = gql`
    mutation CreateExpense($input: ExpenseInput!) {
        createExpense(input: $input) {
            id
        }
    }
`;
