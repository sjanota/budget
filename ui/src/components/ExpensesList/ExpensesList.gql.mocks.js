import { DELETE_EXPENSE, EXPENSES_QUERY } from "./ExpensesList.gql";

export function mockQueryExpenses(expenses) {
  return {
    request: {
      query: EXPENSES_QUERY,
    },
    result: {
      data: {
        expenses
      }
    },
  }
}

export function mockExpensesEvent(event) {
  return {
    result: {
      data: {
        expenseEvents: { ...event, __typename: 'ExpenseEvent' }
      }
    }
  }
}

export function mockDeleteExpense(id) {
  return {
    request: {
      query: DELETE_EXPENSE,
      variables: {
        id: id
      }
    },
    result: jest.fn().mockReturnValueOnce({
      data: {
        deleteExpense: {
          id: id,
          __typename: 'Expense'
        }
      }
    })
  }
}

export const expense1 = {
  id: "5d8265e4d7d8a40795fe1b31",
  title: "Zakupy spożywcze",
  total: {
    integer: 12,
    decimal: 32,
    __typename: 'MoneyAmount'
  },
  location: "Lidl",
  date: null,
  account: null,
  entries: [],
  __typename: "Expense"
};

export const expense2 = {
  id: "5d826618d7d8a40795fe1b33",
  title: "Zakupy spożywcze",
  total: {
    integer: 17,
    decimal: 56,
    __typename: 'MoneyAmount'
  },
  location: "Lidl",
  date: null,
  account: null,
  entries: [],
  __typename: "Expense"
};