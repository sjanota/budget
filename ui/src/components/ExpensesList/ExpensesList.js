import React, {useEffect} from 'react';
import {useQuery} from "@apollo/react-hooks";
import {EXPENSES_EVENTS_SUBSCRIPTION, EXPENSES_QUERY} from "./ExpensesList.gql";
import Table from "react-bootstrap/Table";

function handleExpenseEvent(prev, {subscriptionData}) {
  const event = subscriptionData.data.expenses;
  switch (event.type) {
    case "ADDED": {
      const newExpense = event.expense;
      return {expenses: [...prev.expenses, newExpense]};
    }
    default:
      return prev;
  }
}

function ListHeader() {
  return <thead>
  <tr>
    <th>Tytu≈Ç</th>
    <th>Data</th>
    <th>Suma</th>
    <th>Miejsce</th>
    <th>Konto</th>
  </tr>
  </thead>
}

function ListEntry({expense}) {
  return <tr>
    <td>{expense.title}</td>
    <td>{expense.date}</td>
    <td>{expense.total}</td>
    <td>{expense.location}</td>
    <td>{expense.account && expense.account.name}</td>
  </tr>
}

export default function ExpensesList() {
  const {loading, error, data, subscribeToMore} = useQuery(EXPENSES_QUERY);

  useEffect(() => {
    if (loading) return;
    return subscribeToMore({
      document: EXPENSES_EVENTS_SUBSCRIPTION,
      updateQuery: handleExpenseEvent,
      onError: console.error
    });
  }, [loading, subscribeToMore]);

  if (loading) return <p>Loading...</p>;
  if (error) {
    console.error(error);
    return <p>Error :(</p>;
  }

  return <Table striped bordered hover>
    <ListHeader/>
    <tbody>
    {data.expenses.map(expense =>
      <ListEntry key={expense.id} expense={expense}/>
    )}
    </tbody>
  </Table>
}