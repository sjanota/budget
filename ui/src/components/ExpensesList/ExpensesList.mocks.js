import {EXPENSES_QUERY} from "./ExpensesList.gql";

export function mockQueryExpenses(expenses) {
  return {
    request: {
      query: EXPENSES_QUERY,
    },
    result: {
      data: {
        expenses
      }
    },
  }
}

export function mockExpensesEvent(event) {
  return {
    result: {
      data: {
        expenses: event
      }
    }
  }
}

export const expense1 = {
  id: "5d8265e4d7d8a40795fe1b31",
  title: "Zakupy spożywcze",
  total: 12.32,
  location: "Lidl",
  date: null,
  account: null,
  entries: [],
  __typename: "Expense"
};

export const expense2 = {
  id: "5d826618d7d8a40795fe1b33",
  title: "Zakupy spożywcze",
  total: 12.32,
  location: "Lidl",
  date: null,
  account: null,
  entries: [],
  __typename: "Expense"
};