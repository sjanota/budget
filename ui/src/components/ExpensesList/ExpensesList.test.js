import {mount} from 'enzyme';
import React from 'react';
import {MockedProvider} from "@apollo/react-testing";
import {expense1, expense2, mockExpensesEvent, mockQueryExpenses} from "./ExpensesList.mocks";
import {updateComponent} from "../../testing";
import {createMockLink} from "../../testing/apollo";
import ExpensesList from "./ExpensesList";


it('displays loading initially', async () => {
  const {link} = createMockLink([]);
  console.error = jest.fn();
  const component = mount(
    <MockedProvider link={link}>
      <ExpensesList/>
    </MockedProvider>
  );

  expect(component.find('tbody tr')).toHaveLength(0);
  expect(component.find('p')).toExist();
  expect(component.find('p')).toHaveText("Loading...");
});

it('displays error if occurs', async () => {
  const {link} = createMockLink([]);
  console.error = jest.fn();
  const component = mount(
    <MockedProvider link={link}>
      <ExpensesList/>
    </MockedProvider>
  );
  await updateComponent(component);

  expect(console.error).toBeCalled();
  expect(component.find('tbody tr')).toHaveLength(0);
  expect(component.find('p')).toExist();
  expect(component.find('p').text).toMatchSnapshot();
});

it('displays queried data', async () => {
  const {link} = createMockLink([
    mockQueryExpenses([expense1]),
  ]);
  const component = mount(
    <MockedProvider link={link}>
      <ExpensesList/>
    </MockedProvider>
  );
  await updateComponent(component);

  expect(component.find('tbody tr')).toHaveLength(1);
});

it('updates in ADDED', async () => {
  const {link, sendEvent} = createMockLink([
    mockQueryExpenses([expense1]),
  ]);
  const component = mount(
    <MockedProvider link={link}>
      <ExpensesList/>
    </MockedProvider>
  );
  sendEvent(mockExpensesEvent({type: 'ADDED', expense: expense2, __typename: 'ExpenseAdded'}));
  await updateComponent(component);
  await updateComponent(component);

  expect(component.find('tbody tr')).toHaveLength(2);
});



