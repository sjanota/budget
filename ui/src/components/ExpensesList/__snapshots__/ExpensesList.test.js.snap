// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays error if occurs 1`] = `[Function]`;
